zio-http/src/test/scala/zhttp/service/ClientSpec.scala

```diff
@@ -1,45 +1,69 @@
-      }
-      +
-      testM("empty content") {
-        val actual          = Client.request("http://api.github.com/users/zio/repos", ClientSSLOptions.DefaultSSL)
-        val responseContent = actual.flatMap(_.getBody)
-        assertM(responseContent)(isEmpty)
-      } +
-      testM("text content") {
-        val actual          = Client.request("https://www.google.com", ClientSSLOptions.DefaultSSL)
-        val responseContent = actual.flatMap(_.getBodyAsString)
-        assertM(responseContent)(containsString("Google"))
-      },
-  ).provideCustomLayer(env)
+  val env = ChannelFactory.auto ++ EventLoopGroup.auto() ++ ServerChannelFactory.auto
+
+  val app: HttpApp[Any, Nothing] = Http.collectM[Request] {
+    case Method.GET -> !! / "users" / "zio" / "repos" =>
+      ZIO.succeed(Response.ok)
+
+    case Method.GET -> !! / "users" / "zio" =>
+      ZIO.succeed(Response(status = Status.OK, data = HttpData.fromText("a zio user")))
+
+    case data @ Method.POST -> !! / "users" / "zio" =>
+      data.getBodyAsString
+        .flatMap(content => ZIO.succeed(Response(status = Status.OK, data = HttpData.fromText(content))))
+        .orDie
+
+  }
+
+  override def spec = suiteM("Client")(
+    Server
+      .make(Server.app(app) ++ Server.port(8082))
+      .orDie
+      .as(
+        List(
+          testM("respond Ok") {
+            val actual = Client.request("http://localhost:8082/users/zio/repos", ClientSSLOptions.DefaultSSL)
+            assertM(actual)(anything)
+          } +
+            testM("non empty content") {
+              val actual          = Client.request("http://localhost:8082/users/zio", ClientSSLOptions.DefaultSSL)
+              val responseContent = actual.flatMap(_.getBody)
+              assertM(responseContent)(isNonEmpty)
+            } +
+            testM("POST request expect non empty response content") {
+              val url             = Task.fromEither(URL.fromString("http://localhost:8082/users/zio"))
+              val endpoint        = url.map(u => Method.POST -> u)
+              val headers         = List(Header.userAgent("zio-http test"))
+              val response        = endpoint.flatMap { e =>
+                Client.request(e, headers, HttpData.fromText("test"))
+              }
+              val responseContent = response.flatMap(_.getBody)
+              assertM(responseContent)(isNonEmpty)
+            } +
+            testM("empty content") {
+              val actual          = Client.request("http://localhost:8082/users/zio/repos", ClientSSLOptions.DefaultSSL)
+              val responseContent = actual.flatMap(_.getBody)
+              assertM(responseContent)(isEmpty)
+            } +
+            testM("text content") {
+              val actual          = Client.request("http://localhost:8082/users/zio", ClientSSLOptions.DefaultSSL)
+              val responseContent = actual.flatMap(_.getBodyAsString)
+              assertM(responseContent)(containsString("user"))
+            },
```

Comment:
One of the things we do in ServerSpec is that the `Http` app that we want to test is co-located within the test. This makes it much easier for us to maintain and test. 

Keeping a global HttpApp will become a maintenance mess once the suite becomes large.

Take a look at `AppCollection.scala` to get an understanding of how it's being done.

Reply:
Updated the ClientSpec accordingly